File: node/search/binary_search_rec.js


let array_1 = [...Array(11).keys()];


const binary_search = (search_array, search_value, low=0, high=null) => {
  high = high || search_array.length - 1;
  let middle = Math.ceil((low + high) / 2);
  if (low > high) {
    return -1;
  } else if (search_value == search_array[middle]) {
    return middle;
  } else if (search_value < search_array[middle]) {
    high = middle - 1;
  } else if (search_value > search_array[middle]) {
    low = middle + 1;
  }
  return binary_search(search_array, search_value, low, high);
  // return statement needed to get the result to "trickle down"
};

console.log(binary_search(array_1, 2));
console.log(binary_search(array_1, 2.5));

File: python/search/binary_search_rec.py
from math import ceil

list_1 = [1,23,4,5,6,7,8,9,10]



def recursive_search(search_list, search_value, left_i=0, right_i=None):
    right_i = right_i or len(search_list)-1
    middle_i = ceil(sum(left_i, right_i) / 2 )
    if left_i > right_i:
        return -1
    elif search_value == search_list[middle_i]:
        return middle_i
    elif search_value > search_list[middle_i]:
        left_i = middle_i + 1
    elif search_value < search_list[middle_i]:
        right_i = middle_i - 1
    recursive_search(search_list, search_value, left_i=left_i, right_i=right_i)

